Add ability to use direct urls from cover services in UI

xmorave2 opened the pull request:
We want to use local book covers service as CDN, so we need to pass the url provided by service in directly as image source in generated html.

TODO, test covers on these templates:

- [x] RecordDriver/DefaultRecord/collection-info.phtml
- [x] RecordDriver/DefaultRecord/core.phtml
- [x] RecordDriver/EDS/core.phtml
- [x] RecordTab/similaritemscarousel.phtml
- [x] search/results.phtml 
- [x] Add content security policy notes to ObakyKnih configuration after #1494 is merged

demiankatz commented:
Thanks, @xmorave2, this is interesting... Did you follow the conversation a few years ago on [VUFIND-1140](https://vufind.org/jira/browse/VUFIND-1140), which discussed having an AJAX API for loading cover images? I wonder if it would be worth taking that approach here, in order to have the ability to implement a mix of proxied and unproxied images if necessary (i.e. rather than having a global setting for which behavior to use, instead have each cover provider report its preference). That would also make it possible to pass through captions, copyright messages, and other things that we currently can't do.

Perhaps that's too complicated and beyond the scope of what you need... but I wonder if this is a good opportunity to finally get that work done. I'm willing to help with the development if you think it's worth the effort (and the extra time before this can get merged). What do you think?

demiankatz commented:
@xmorave2, this hasn't seen any activity in a few weeks; just thought I'd touch base and see if it slipped through the cracks. No rush, of course!

xmorave2 commented:
@demiankatz I am going to work again on our covers project in the near future. I should first read the VUFIND-1140 report thouroughly and think it out together with our needs.

xmorave2 commented:
I read through the discussion on #507 and VUFIND-1140.

My conclusion from this is:

- We need an ajax handler, which is able to send back url (direct or proxied, it depends) and some other data related to cover (link to service, terms of usage, ...)
- This way should be an alternative to current implementation
- The ajax handler should be able to handle more then one cover at once (e.g. search results list)
- Some cache of cover availability on specific services should be implemented too

Do I understand it correctly?

And, I have another idea related to content plugins:
We couuld pass whole record to content plugins (Covers, TOC, Excerpts, ...) - we are using only ISBNs on all of them (except covers), but some services may (and ObalkyKnih.cz does) accept also other identifiers. This way, the plugin could take the identifiers it needs, regardless of VuFind capabilities to pass all possible identifiers.

demiankatz commented:
Yes, you understand correctly.

Regarding your idea of passing the whole record for content, it looks like this should be a fairly easy change to make for most of the content code; the existing loadByIsbn() method could be retained for backward compatibility, but a loadByFullRecord() could be added and used instead for most cases. The reason we started out relying on ISBNs alone is because for cover loading, we had to pass identifiers through GET parameters, and we didn't want to have to do extra Solr lookups for cover generation... but for all of the other content, we already have access to the full record, so there's no reason not to use it.

Please let me know if I can be of any assistance with any of this work; I know it's a fairly large project!

xmorave2 commented:
Thanks @demiankatz! I will start with it, and ask for a help eventually if needed ;)

xmorave2 commented:
I worked a bit on covers project.

This is not finished yet, but I want to share what I have done.

There are two main changes:
- refactoring of direct urls - adds ability to define if use proxied or direct url on provider level
- added plugin for our local provider ObalkyKnih.cz (which means "book covers" translated to english)

demiankatz commented on config/vufind/obalkyknih.ini:
```diff
@@ -0,0 +1,36 @@
+; Note: You have to register an account on obalkyknih.cz and configure your VuFind's instance IP address/referrer url
+
+[ObalkyKnih]
+base_url1 = https://cache.obalkyknih.cz
```
What is the purpose of `base_url1`/`base_url2`? Would it be better to implement this as `base_url[]` so an arbitrary number could be configured as needed?

xmorave2 replied:
ObalkyKnih service has two servers, when one is down, api client could use the second one as fallback (not implemented in my code yet).

But you are right, that array would be better here, I'll fix it.

xmorave2 replied:
Done just now

demiankatz commented:
Thanks, @xmorave2; at a glance this looks like a great start. I'll give it a closer review in the next few days when I have a bit more time. In the meantime, I just had time to ask one quick minor question.

demiankatz commented on module/VuFind/src/VuFind/AjaxHandler/GetRecordCover.php:
```diff
@@ -0,0 +1,128 @@
+     */
+    public function handleRequest(Params $params)
+    {
+        $recordId = $params->fromQuery('recordid');
```
You're going to want to pass the record source (i.e. the `getSourceIdentifier()` value from the record driver) through as well so it can be passed to the second parameter of load() below, or else this will only work for Solr records and not records from other backends.

demiankatz commented on module/VuFind/src/VuFind/AjaxHandler/GetRecordCover.php:
```diff
@@ -0,0 +1,128 @@
+        }
+
+        $cover = $record->tryMethod('getThumbnail');
+        if ($cover === null) {
```
I'm not sure I understand the purpose of this logic. Why are thumbnails being blocked if the `getThumbnail()` method is undefined?

demiankatz commented on module/VuFind/src/VuFind/View/Helper/Root/Record.php:
```diff
@@ -68,6 +72,13 @@ class Record extends AbstractClassBasedTemplateRenderer
+     *
+     * @var CoverLoader
+     */
+    protected $coverLoader;
```
Since this is not being injected through the constructor, it should probably be initialized to null, and safety checks put in place to avoid trying to call methods on it if it is not populated correctly... or, if practical, maybe it should be refactored to a constructor parameter to guarantee its presence.

demiankatz commented on module/VuFind/src/VuFind/View/Helper/Root/Record.php:
```diff
@@ -570,9 +593,35 @@ public function getQrCode($context, $extra = [], $level = "L", $size = 3,
-        return $this->coverRouter
-            ? $this->coverRouter->getUrl($this->driver, $size)
-            : false;
+        $handlers = $this->coverLoader->getHandlers();
```
This refactoring seems to be moving a lot of work from the cover router into the record view helper. Might it be preferable to pass additional parameters into the router to be able to do the work there? (I haven't reviewed this closely, so maybe there is a strong reason to do it this way... but in general, my preference would be to keep the view helper thinner and delegate work to a more specialized class, if practical).

xmorave2 replied:
I moved this logic to cover router, it does make more sense here. Thanks.

demiankatz commented on module/VuFind/src/VuFind/AjaxHandler/GetRecordCover.php:
```diff
@@ -0,0 +1,128 @@
+<?php
+/**
+ * GetRecordCover AJAX handler.
```
Is this handler being used yet, or just work in progress? I couldn't spot a place where it comes into play yet, but maybe I wasn't looking closely enough.

xmorave2 replied:
Ajax handler is WIP, not finished yet, I am going to work on it as a next step.

demiankatz replied:
Sounds good -- just let me know whenever you're ready for another review!

xmorave2 commented:
@demiankatz Could you look at what I have now?
I refactored the logic of getting cover url from view helper to cover router
I added ajax handling - javascript and template changes.

Now I am going to work on tests

demiankatz commented:
Thanks, @xmorave2; I'll take another look as soon as I can, though it may take a couple of days. It's our end of fiscal year at Villanova, so I may be unusually busy for a little while.

xmorave2 commented:
no problem @demiankatz, it could wait ;)

demiankatz commented on module/VuFind/src/VuFind/AjaxHandler/AjaxHandlerInterface.php:
```diff
@@ -44,6 +44,7 @@ interface AjaxHandlerInterface
     const STATUS_HTTP_BAD_REQUEST = 400; // bad request
     const STATUS_HTTP_NEED_AUTH = 401;   // must login first
     const STATUS_HTTP_FORBIDDEN = 403;   // method is unavailable
+    const STATUS_HTTP_NOT_FOUND = 404;   // object/data not found
```
This new constant does not appear to be used anywhere; I don't object to adding it, but just curious if something is missing.

demiankatz commented on module/VuFind/src/VuFind/Cover/Router.php:
```diff
@@ -76,12 +88,38 @@ public function getUrl(RecordDriver $driver, $size = 'small')
             return false;
```
I understand now why you were failing out of the AJAX method when getThumbnail returned an empty result -- it was the parallel to this logic. We should probably make the logic align in both places, but at this point I'm not sure whether it's better to remove this logic or add it back to the AJAX handler. (And apologies for any confusion I'm causing -- it's been a while since I refreshed my memory on how this all works).

demiankatz replied:
I've reviewed the logic again, and it looks fine to me now; resolving this thread.

demiankatz commented on module/VuFind/src/VuFind/View/Helper/Root/AjaxCovers.php:
```diff
@@ -0,0 +1,80 @@
+class AjaxCovers extends \Laminas\View\Helper\AbstractHelper
+{
+    /**
+     * Render javascript for load cover by ajax for detial
```
Typo ("detial") -- also repeated below.

xmorave2 replied:
I'll fix it

demiankatz commented on themes/bootstrap3/templates/ajax/resultgooglemapinfo.phtml:
```diff
@@ -7,6 +7,7 @@
       <div class="mapInfoResult <?php if ($i % 2 == 0): ?>alt <?php endif; ?>record<?=$i ?>">
         <div class="mapInfoResultThumb">
           <?php if ($thumb = $this->record($record)->getThumbnail()): ?><img class="mapInfoResultThumbImg" src="<?=$this->escapeHtmlAttr($thumb) ?>"/><?php endif; ?>
+          <?=$this->ajaxCovers()->detail('small')?>
```
I see that you've added ajaxCovers support to a lot of templates where thumbnails are rendered, and I imagine that some of these may be difficult to test without a fairly specific setup. It might be worthwhile to create a set of TODO checkboxes for testing some of the narrower edge cases. If there's anything you can't easily check yourself, please make a TODO box and I'll be happy to help with the testing.

xmorave2 replied:
I added todo list for templates and will try to test them

demiankatz replied:
Thanks! Let me know if you need help and I'll take a closer look when I have a bit more time available.

xmorave2 replied:
I'll try to test what I will be able and I want also create some unit test for ajax handler, after that, i will need you next review round

demiankatz replied:
Thanks, that sounds great; just let me know once you are ready.

demiankatz commented on module/VuFind/src/VuFind/Cover/Loader.php:
```diff
@@ -318,7 +319,7 @@ public function loadImage($settings = [])
         // in as an array:
         $settings = is_array($settings)
-            ? array_merge($this->getDefaultSettings(), $settings)
+            ? $settings
```
By factoring out the `getDefaultSettings()` merge here, we are now unable to guarantee that `$settings['title']`, `$settings['author']`, and `$settings['callnumber']` will be set below, which could trigger notices. It might be worth adding some `?? null` fallbacks to prevent that problem.

xmorave2 replied:
`getDefaultSettings` is called from `storeSanitizedSettings` (which is called just few lines bellow), as you can't load image without calling storeSanitizedSettings, it should be OK I think...

demiankatz replied:
Ahh, yes, I see. I think I was confused because `storeSanitizedSettings` was changed to public. Is that actually necessary? If it still serves only as a support method for `loadImage`, it would probably be better to change it back to protected.

xmorave2 replied:
`storeSanitizedSettings` is needed in cover router because the ability to get direct urls, so we need it to be public...

demiankatz replied:
Ahh, yes, I see. I think it would perhaps be less confusing if we could refactor the code slightly to avoid the need to call `storeSanitizedSettings` separately. That just feels to me like we're exposing too many class internals in public. How would you feel about adding:

<pre>
public function getIdentifiersForSettings($settings)
{
    $this->storeSanitizedSettings($settings);
    return $this->getIdentifiers();
}
</pre>

...instead of making `getIdentifiers` and `storeSanitizedSettings` public? I think that might make it easier to refactor the code in future if we want to make it more functional and less dependent on side effects.

Sorry if I'm being too picky here; this just feels like it might clarify things slightly.

xmorave2 replied:
Thanks @demiankatz, this does make sense, I've refactored it that way.

demiankatz replied:
Great, thanks!

demiankatz commented on module/VuFind/src/VuFind/Cover/Loader.php:
```diff
@@ -435,31 +436,14 @@ protected function fetchFromAPI()
+            $urls = $this->getCoverUrls();
+            foreach ($urls as $url) {
+                $success = $this->processImageURLForSource(
+                    $url['url'], $url['handler']->isCacheAllowed, $url['apiName']
```
I think this should be `isCacheAllowed()`, not just `isCacheAllowed`.

xmorave2 replied:
You are right, I'll fix it

demiankatz commented on module/VuFind/src/VuFind/Cover/Loader.php:
```diff
@@ -672,4 +656,60 @@ protected function processImageURL($url, $cache = true)
+                }
+            } catch (\Exception $e) {
+                $this->debug(
+                    get_class($e) . ' during processing of ' . $apiName
```
`$apiName` is not defined; should be `$handler['apiName']`.

xmorave2 replied:
Right, I'll fix it

demiankatz commented:
Thanks again for the progress, @xmorave2. I'll try to do some hands-on testing of this next week; I'll be sure to cover the EDS templates while I'm doing that, as they're a bit unusual.

demiankatz commented on module/VuFind/src/VuFind/Cover/Router.php:
```diff
@@ -76,12 +88,37 @@ public function getUrl(RecordDriver $driver, $size = 'small')
         }
 
-        // Array?  It's parameters to send to the cover generator:
+        // Array? It's parameters to send to the cover generator:
```
Should we reverse the logic here, so if `$thumb` is not an array, we return it immediately? Otherwise, I think we could potentially override a URL specified by the driver as a string with a direct URL fetched from the cover services, which would be a change in behavior. Currently, a string coming from getThumbnail always overrides dynamic lookups.

demiankatz replied:
It looks like this older comment may still need to be addressed. If you like, I can set up and test some scenarios, and make adjustments if necessary; just curious if you have any opinions before I do.

demiankatz replied:
Never mind, I see that you did address this after all. :-)

demiankatz commented on themes/bootstrap3/less/components/search.less:
```diff
@@ -151,7 +151,7 @@ header .container.navbar { margin-bottom: 0; }
       &.large img {
         width: auto;
-        max-width: 100%;
+        max-width: initial;
```
What is the intent behind this change? It is causing problems for me:

![image](https://user-images.githubusercontent.com/309069/85715489-82547700-b6b9-11ea-9ad9-f4bd61f1a366.png)

demiankatz replied:
(I should note that this record displays correctly when AJAX covers are enabled, but it displays incorrectly as above when the setting is disabled).

xmorave2 replied:
I probably had to make some mistake, but not sure where. As you mentioned - it works with ajax covers, and that's why I changed it. But i forget to test it without ajax covers...

demiankatz replied:
We can probably ask for @crhallberg's assistance on this if necessary; he might have some ideas.

xmorave2 replied:
I tried to fix it, could you please test it? It is in commit 2a6b4f4

demiankatz replied:
Looks good to me!

demiankatz commented on themes/bootstrap3/templates/RecordTab/similaritemscarousel.phtml:
```diff
@@ -17,8 +17,7 @@
             <a class="hover-overlay" href="<?=$this->recordLink()->getUrl($data)?>">
-              <?php $thumb = $this->record($data)->getThumbnail('large'); ?>
-              <img src="<?=$thumb ?>" title="<?=$data->getTitle() ?>"/>
+              <?=$this->record($data)->getCoverDetails('core', 'large')['html']?>
```
This was not displaying correctly in the previous version; I think I've fixed it by simplifying the code, but please let me know if I've failed to account for anything.

xmorave2 replied:
It's loooking good. Thank!

demiankatz commented on themes/bootstrap3/templates/RecordTab/similaritemscarousel.phtml:
```diff
@@ -53,6 +52,7 @@
 <?php endif; ?>
 <?php
   $script = <<<JS
+loadCovers();
```
This throws an undefined function error when AJAX covers are disabled. I think we need to check if it's defined before we call it. Also, is there a risk that this will cause some images on the page to double-load, because the function was previously called through the document ready event? If there's not already a mechanism in place to protect against double-loading, we might want to add one (perhaps a data attribute on the image, or something similar).

xmorave2 replied:
As I remember, it needed to be called secondly when a tab is loaded by ajax... But I add a check if ajax covers are enabled.

 I agree we should prevent double loading of images, I'll fix it, but probably not till monday...

xmorave2 replied:
I've added check for ajaxCoversEnabled in similar items tab

demiankatz replied:
Thanks!

demiankatz commented on themes/bootstrap3/templates/record/cover.phtml:
```diff
@@ -5,5 +5,5 @@
   <?php if ($this->link && !empty($alt)): ?></a><?php endif; ?>
 <?php else: ?>
-  <img src="<?=$this->url('cover-unavailable')?>" <?php if ($linkPreview): ?>data-linkpreview="true" <?php endif; ?>class="recordcover" alt="" />
+  <img src="<?=$this->url('cover-unavailable')?>" <?php if ($linkPreview): ?>data-linkpreview="true" <?php endif; ?>data-recordsource="<?=$driver->getSourceIdentifier()?>" data-recordid="<?=$driver->getUniqueID()?>" data-coversize="<?=$size?>" class="recordcover" alt="" />
```
It's a little disconcerting that images display as unavailable before popping in with AJAX. Do we need to differentiate between "loading" status and "unavailable" status? (If that's too much hassle to deal with right now, I can live with it as-is in the interest of getting the feature merged, but I think we should at least open a JIRA ticket to make that a future enhancement to the system).

xmorave2 replied:
I implemented a simple spinner here now

demiankatz replied:
Thanks! It wasn't quite working for me; for some reason, the paths in compiled.css were wrong. I ran `grunt less` again and it fixed everything. How are you compiling your LESS? I wonder if there is a bug in the process somewhere, or some inconsistency between our environments.

I also changed the text to simply "Loading" so we can use existing translations without having to create a new cover-specific one.

xmorave2 replied:
I am compiling into CSS using util/cssBuilder command in Vagrant box. Then it worked for me on the same box as server.

demiankatz replied:
I think that the PHP compiler sometimes produces less portable code than the Javascript-based compiler. I'd suggest using `grunt less` instead of util/cssBuilder if practical... but if that's not possible, it's not a problem for me to recompile on my end as needed. :-)

xmorave2 replied:
thanks @demiankatz I'll try it next time ;)

demiankatz commented on themes/bootstrap3/templates/RecordDriver/EDS/core.phtml:
```diff
@@ -7,12 +7,14 @@
-    <?php else: ?>
+    <?php if ($cover): ?>
+      <?=$cover?>
+    <?php else: //TODO an icon as default cover? ?>
```
This definitely does not work correctly; the icons never display because getCoverDetails always displays the unavailable image. Also note that there's an EDS/result-list.phtml that also displays icons, but that has not been modified in this PR, and it still seems to work correctly without the ajaxcovers setting, but to never load covers at all when ajaxcovers is enabled.

Would it make sense to provide a mechanism for alternate content to display when a cover is unavailable? This might tie in with the idea of "loading" vs. "unavailable" images -- when we're in AJAX mode, we can display a "loading" image until the AJAX handler is called, then we either replace the loading image with the real content, or else we reveal a hidden element with either the "unavailable" image (in most situations) or the content icon (for EDS).

If that's too complicated, we at least need to figure out a way to disable the AJAX functionality within the context of EDS so that it doesn't break the existing templates.

Any thoughts/preferences?

xmorave2 replied:
I do agree, that such concept of differantiation loading/unavailable does make sense a could help us to fix the EDS icons/covers... 

I do like the idea, so I'll try to implement it on this PR and will see, how it will go

demiankatz replied:
Sounds good! I'll focus on the other open PR's tomorrow to see if I can get them closer to completion, and I'll probably have some time to help with this next week if necessary.

demiankatz replied:
@xmorave2, the EDS search results still work correctly with ajax covers disabled, and fail to display any covers (instead showing icons only) with ajax covers enabled.

On the record view, items that have cover images now work correctly with or without ajax covers enabled. However, items that have no cover image are now broken regardless of how the ajax cover setting is configured -- they just show a "Loading..." status that never loads. 

I realize that this is difficult for you to test and develop, and I'm willing to put some time into it... but before I do, I wanted to find out whether you expected this to be working correctly, or if you had any suggestions for which logic I should try to port into the EDS templates.

xmorave2 replied:
I hope I finally fixed the EDS covers, but I was not able to fully test it. Could you look at it @demiankatz? Thanks!

demiankatz replied:
Thanks, I should have time to review later today; I'll let you know how it goes!

demiankatz replied:
It took minor adjusting to get it working, but I think it's good now!

demiankatz commented:
@xmorave2, I had time to make more progress on this today; see all of the comments above -- there are still a few issues we need to figure out. I can probably help with some of them next week, but I'm out of time for today. As always, thanks for your help; we're definitely getting closer to a solution!

demiankatz commented:
@xmorave2, we're making more progress -- I've left updated comments on a few open threads. I'm going to move on to reviewing other open pull requests for today, but I'll have more time tomorrow if we're ready to make further progress!

demiankatz commented:
@xmorave2, I spent a bit more time on this today. I'm having trouble getting ObalkyKnih to work; I haven't managed to get any cover URLs yet. However, in the process of testing this, I discovered a bug, which I believe I have fixed in https://github.com/vufind-org/vufind/pull/1566/commits/3287eaa23d69127db8bbbf64422f1e853d9324de -- since we rely on null coalescing (`??`) to pick a URL at the end of the function, we can never set $directUrl to false, because that qualifies as a non-null value, and ends up overriding the indirect URL! Hopefully the fix makes sense.

I'm out of time for today, but I may have some more time tomorrow. If you have time to answer my EDS comments from last week and/or troubleshoot my ObalkyKnih issues, that would be great. If not, we may have to wait to merge this until after the 7.0 release; time is beginning to run out.

demiankatz commented:
@xmorave2, update: I realized I had my referer setting wrong in the .ini file, and when I fixed that, I was able to get ObalkyKnih working, so I have confirmed that my fix seems to be working correctly... so I think that just leaves us with the task of figuring out the EDS problem. If you have any further thoughts on that, please let me know; if not, I'll dig deeper tomorrow.

Also, one very minor problem: I notice that cover images (including the "no cover available" image) are getting upscaled in the search results when they're below a certain threshold, which can look a bit ugly. Is this intentional, or an accidental side effect of some of the LESS changes?

xmorave2 commented:
> Also, one very minor problem: I notice that cover images (including the "no cover available" image) are getting upscaled in the search results when they're below a certain threshold, which can look a bit ugly. Is this intentional, or an accidental side effect of some of the LESS changes?

That happens by accident, but I am not sure what is the cause of this... @crhallberg, could you please look at it when you have time?

xmorave2 commented:
Thanks @demiankatz for the bunch of commits you done here! I appreciate it. I added a one more to (hopefully) fix covers in EDS. I refactored record/cover.phtml template to be record driver specific, to have ability to fallback to icons for EDS.

crhallberg commented:
re:upscaling / There is an issue with the LESS but not your changes. Also, as far as I can tell this bug only affects the mobile view. I have a solution to the issue but it may be best to wait until after this PR.

demiankatz commented:
@xmorave2, I think we may be ready to merge, but it's been an extremely long day, and I don't want to overlook something while I'm tired, so I'll try to review this one more time tomorrow. If you and/or @crhallberg don't mind reviewing my latest changes, I'd be most grateful!

A few notes:

1.) The EDS icons weren't centering properly, so I added a `margin: auto` to fix them.
2.) The "Loading..." icon never went away on EDS icons when AJAX covers were disabled, so I moved it into the appropriate conditional area.
3.) When AJAX covers were enabled, they were overflowing the text on the record page (similar to the problem I reported before). This was related to the `max-width: initial` you added to search.less. I solved the problem by simply removing `.record` from the selector. I don't really understand what the purpose of this style is, though, so I'm concerned this may not be the right solution... or maybe the whole thing is unnecessary. This is the area that could probably use the most careful review.

Anyway, thanks again; I think we are close! More tomorrow.

xmorave2 commented:
@demiankatz: Thanks for your fixes, they looks good to me. Removing `.record` does not seen to broke anything for me, so I am OK with that

demiankatz commented:
@xmorave2, the EDS code revealed an inefficiency: when the record driver returns a hard-coded thumbnail URL instead of an array of dynamic parameters, there's no reason to go through the AJAX process to resolve anything -- it just creates pointless overhead. I've refactored slightly to short-circuit in this case: https://github.com/vufind-org/vufind/pull/1566/commits/27937ff4984e5a94aca037b0078dcfade7bcf1b4 -- seems to be working (and I'll continue to test), but please let me know if you have any concerns about this change.

demiankatz commented:
@xmorave2, another significant brainstorm: when there is no thumbnail data available at all, that's another situation where we should short-circuit the AJAX process. I adjusted the cover router to differentiate between a false return (no cover) and a null return (AJAX check needed), which then allowed me to make the two versions of cover.phtml more consistent, making the EDS-specific logic more distinct. See commit https://github.com/vufind-org/vufind/pull/1566/commits/37178ef577c89f5637370de8b6fe0357ba35dfbf -- once again, feedback/double-checks welcomed!

demiankatz commented:
...and one more adjustment: links around covers were not working properly in AJAX mode; should be fixed now.

demiankatz commented:
...and it's merged! Thanks again, @xmorave2, for all the work on this. If you have outstanding concerns, please feel free to open a new PR for additional adjustments.

xmorave2 commented:
@demiankatz Thanks for your follow-ups! I've just tested the current master and it's looking good. Thanks again for all your help on this!

EreMaijala commented:
I'm late to the game, but would you oppose to this being expanded in the future a bit to allow multiple images? We have those especially for non-library materials, and out-of-the-box support should make things easier here.

xmorave2 commented:
> I'm late to the game, but would you oppose to this being expanded in the future a bit to allow multiple images? We have those especially for non-library materials, and out-of-the-box support should make things easier here.

@EreMaijala: I have no objection against multiple images. I agree it could be useful. But it is not the thing I am going to work on in the near future...

EreMaijala commented:
@xmorave2 I didn't expect you to. And there's of course the question of how those are displayed. In any case this foundation should make things much easier for us. Thanks for your work!

demiankatz commented:
@EreMaijala, I agree that support for multiple images would be useful in the future -- it could potentially improve performance and reduce load on Solr and other services by reducing the number of calls when AJAX is in use. Another area of future expansion is the inclusion of a mechanism in the API for displaying text alongside the images, so that we can add support for cover services that require some form of attribution alongside the images. As you say, the goal for now was to create a foundation that we could build upon, and I think we've made a great start, but hopefully we can add these other things in a future release as demand for them increases. I always welcome new PRs to continue the work!
